src/qibojit/custom_operators/backends.py

```diff
@@ -384,3 +383,281 @@ def swap_pieces(self, piece0, piece1, new_global, nlocal):
+        else:
+            ncontrols = 0
+            controls = self.np.empty()
+        adjoint = 0
+
+        state = self.cast(state)
+        gate = self.cast(gate)
+        assert state.dtype == gate.dtype
+        data_type, compute_type = self.get_cuda_type(state.dtype)
+        if isinstance(gate, self.cp.ndarray):
+            gate_ptr = gate.data.ptr
+        elif isinstance(gate, self.np.ndarray):
+            gate_ptr = gate.ctypes.data
+        else:
+            raise ValueError
+
+        workspaceSize = self.cusv.apply_matrix_buffer_size(handle,
+                                                           data_type,
+                                                           nqubits,
+                                                           gate_ptr,
+                                                           data_type,
+                                                           self.cusv.MatrixLayout.ROW,
+                                                           adjoint,
+                                                           ntarget,
+                                                           ncontrols,
+                                                           compute_type
+                                                           )
+
+        # check the size of external workspace
+        if workspaceSize > 0:
+            workspace = self.cp.cuda.memory.alloc(workspaceSize)
+            workspace_ptr = workspace.ptr
+        else:
+            workspace_ptr = 0
+
+        self.cusv.apply_matrix(handle,
+                               state.data.ptr,
+                               data_type,
+                               nqubits,
+                               gate_ptr,
+                               data_type,
+                               self.cusv.MatrixLayout.ROW,
+                               adjoint,
+                               target.ctypes.data,
+                               ntarget,
+                               controls.ctypes.data,
+                               ncontrols,
+                               0,
+                               compute_type,
+                               workspace_ptr,
+                               workspaceSize
+                               )
+
+        self.cusv.destroy(handle)
+        return state
+
+    def two_qubit_base(self, state, nqubits, target1, target2, kernel, qubits=None, gate=None):
+        # initialize cuStateVec library
+        handle = self.cusv.create()
+
+        ntarget = 2
+        target1 = nqubits - target1 - 1
+        target2 = nqubits - target2 - 1
+        target = self.np.asarray([target2, target1], dtype=self.np.int32)
+        ncontrols = len(qubits) - 2 if qubits is not None else 0
+        controls = self.np.asarray([i for i in qubits.get() if i not in [target1, target2]], dtype = self.np.int32)
+        adjoint = 0
+
+        state = self.cast(state)
+        gate = self.cast(gate)
```

Comment:
This will also fail if `gate` is `None`. Again this never happens in the code, that's why it's not captured in tests. If that is the case we need to decide whether `gate` can be `None` here and treat this case properly, or alternatively make `gate` a required argument for this method.